pgxn/neon/pagestore_smgr.c

```diff
@@ -613,6 +613,14 @@ prefetch_on_ps_disconnect(void)
 		Assert(slot->status == PRFS_REQUESTED);
 		Assert(slot->my_ring_index == ring_index);
 
+		/*
+		 * Drop connection to all shards which have prefetch requests.
+		 * It is not a problem to call disconnect multiple times on the same connection
+		 * because disconnect implementation in libpagestore.c will check if connection
+		 * is alive and do nothing of connection was already dropped.
+		 */
```

Comment:
> Can't we "just" wait for responses on other connections, or reset prefetches for only the dropped connection?

Prefetch can take arbitrary long time. Right now `efficient_io_conciurrency` can set up to 100 and waiting 100 pages may take quite long time.

Also please notice that this error is detected in low level - SMGR. It is not so good idea to block backend for a long time here - (pinned shared buffer, hold locks, ...).

Certainly it it not necessary to wait for all responses in error handler. We can implement some state machine and consume responses lazily. But it significantly complicates all pageserver logic and make it more error prone. 
Assume that we have send prefetch requests to shards 1,2,1,1,3,1,2,1,2,3,3
Nope connection with shard 1 is lost. We try to reestablish it and succeed.  If we do not want to drop connections to shards 2 and 3, then we should somehow ignore this existed prefetch requests to 1 but allow to place new prefetch requests to 1 in ring buffer! I do not want to say that it is impossible, we can use "generations", we can mark all existed requests to 1 as invalid,...

But right now I prefer to provide much simpler and reliable solution: just reset all connection and cleanup prefetch queue. I still do not think that PS restart will happen frequently, even if we have multiple shards. But my key point is that trying to preserve part of prefetch queue doesn't cost this efforts: prefetch is used only for OLAp and default prefetch distance is very small (1?).

Reply:
I still think that PS crash happens very rarely (once per year?) and planned restart of PS (because of new deploy) most likely will affect all shards.
Ny main concerned is increased complexity of prefetch logic caused by presence in prefetch ring both live and dead  prefetch requests. We should somehow ignore last when waiting for response or when checking if such prefetch request was already issued.  As a consequence we may have in prefetch ring two requests with the same tag, but one of them should be skipped. Not sure ifs current prefetch implementation will correctly handle it. 

So unless you insist, I prefetch to merge this simplified fix rather than try to address all this issues.

Reply:
Ok, I agree. But still it will happen not more frequently than once per day.